### PR TITLE
refactor(arcgis): overhaul of arcgis connector and converter patterns to enable data extraction

### PR DESCRIPTION
**Pull request by clairekuang**
## Description & motivation
This pr is a big refactor of the arcgis connector to enable data extraction workflows (sending `ArcgisObject: IDataObject`). It includes the following changes to align arcgis with all other data extractor patterns:

- Objects of `CoreObjectBase` type are now passed to the single `TopLevelConverter`. Previously, this was a mixture of objevcts and layers
- All layer logic is handled in the connector. Previously, `LasDatasetLayer`s were passed to the converter. These are now passing individual points to the converter, instead of sending the layer as a pointcloud
- `DisplayValueExtractor` and `PropertiesExtractor` classes added to the converter to assist in object creation in the top level converter.
- `ColorUnpacker` class added to connector to handle color proxy creation during mapmember iteration
- CRS information is added to the root collection, instead of to every single layer

Some async logic was refactored in the connector to improve safety.
Some poor conversion logic in the converters was removed to improve performance.
Reduced complexity in root object builder, removed pragma warning disable.

Note: receive needs an equally large refactor to handle `DataObject`s only. This pr will disable the receive binding until that refactor is completed. Metrics from qgis v2 show that sending is much more common than receiving (~2500 incidents vs 220). Arcgis currently has 1 MAU.

## Changes:

- arcgis converter and connector

## To-do before merge:

- [x] Test sending
- [x] disable receiving
- [ ] add send caching

## Screenshots:

In arcgis file:
{FFD12B3F-01CA-4259-9CE9-D54A4AB3B829}

With this pr:
{73FAAA63-AA1D-464D-BA4D-BD0798042B93}
https://latest.speckle.systems/projects/2295cb26a0/models/874439f06c

## Validation of changes:
Tested sending the sample arcgis file. Note: our pro license does *not* include 3d extensions, meaning I couldn't test point clouds.

### CODE REVIEW

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/Extensions/SpeckleApplicationIdExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+namespace Speckle.Converters.ArcGIS3.Extensions;
+
+public static class SpeckleApplicationIdExtensions
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/Extensions/SpeckleApplicationIdExtensions.cs:3]: i like these classes over host apps which application ids are mystery

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToHost/Raw/GeometryToHostConverter.cs [hunk 2]
```diff
@@ -8,17 +8,14 @@ public class GeometryToHostConverter : ITypedConverter<IReadOnlyList<Base>, ACG.
 {
   private readonly ITypedConverter<List<SOG.Polyline>, ACG.Polyline> _polylineConverter;
   private readonly ITypedConverter<List<SOG.Point>, ACG.Multipoint> _multipointConverter;
-  private readonly ITypedConverter<List<SOG.Polygon>, ACG.Polygon> _polygonConverter;
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToHost/Raw/GeometryToHostConverter.cs:11]: out of curiosity what is the difference Polygon and Polyline apart from `isClosed` prop

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/DisplayValueExtractor.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      case ACD.Raster.Raster raster:
+        yield return _gisRasterConverter.Convert(raster);
+        break;
+
+      case ACD.Analyst3D.LasPoint point:
+        yield return _pointConverter.Convert(point.ToMapPoint());
+        break;
+
+      default:
+        yield break;
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/DisplayValueExtractor.cs:53]: should we log?

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/DisplayValueExtractor.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    if (geometryField is null)
+    {
+      yield break;
+    }
+
+    var shape = (ACG.Geometry)row[geometryField];
+    switch (shape)
+    {
+      case ACG.MapPoint point:
+        yield return _pointConverter.Convert(point);
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/DisplayValueExtractor.cs:75]: do really `pointConverter` returns list of `Base`s?

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/PropertiesExtractor.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
+namespace Speckle.Converters.ArcGIS3.ToSpeckle.Helpers;
+
+public sealed class PropertiesExtractor
+{
+  public PropertiesExtractor() { }
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/PropertiesExtractor.cs:5]: can be removed

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/PropertiesExtractor.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    switch (coreObjectsBase)
+    {
+      case ACD.Row row:
+        return GetRowFields(row);
+    }
+
+    return new();
+  }
+
+  public Dictionary<string, object?> GetRowFields(ACD.Row row)
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/PropertiesExtractor.cs:18]: can be private?

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/PropertiesExtractor.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  {
+    Dictionary<string, object?> rowFields = new();
+    foreach (ACD.Field field in row.GetFields())
+    {
+      // POC: do not set null values
+      // POC: we are not filtering by the layer visible fields
+      if (FieldValueToSpeckle(row, field) is object value)
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Helpers/PropertiesExtractor.cs:25]: even if it returns null, i guess it is still `object`? below statement might be better
```c#
var value = FieldValueToSpeckle(row, field);
if (value != null)
{
    rowFields[field.Name] = value;
}
```

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/MultipatchFeatureToSpeckleConverter.cs [hunk 8]
```diff
@@ -44,83 +42,127 @@ public IReadOnlyList<Base> Convert(ACG.Multipatch target)
...
+
+  private SOG.Mesh GetMeshFromTriangleStripPatch(List<ACG.MapPoint> points)
+  {
+    List<double> pointCoords = points.SelectMany(x => new List<double>() { x.X, x.Y, x.Z }).ToList();
+    List<int> faces = new();
+    List<double> vertices = new();
+
+    for (int i = 0; i < points.Count; i++)
+    {
+      if (i >= 2) // every new point adds a triangle
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/MultipatchFeatureToSpeckleConverter.cs:93]: this assumption smells a bit. i guess we rely on order of `points` here to create proper triangles. even if it ordered OK, i wonder what happens if points in non-convex shape, we might have overlapping triangles. (I know this was pre-existing😒)

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/MultipatchFeatureToSpeckleConverter.cs [hunk 9]
```diff
@@ -44,83 +42,127 @@ public IReadOnlyList<Base> Convert(ACG.Multipatch target)
...
+
+  private SOG.Mesh GetMeshFromTrianglesPatch(List<ACG.MapPoint> points)
+  {
+    List<double> pointCoords = points.SelectMany(x => new List<double>() { x.X, x.Y, x.Z }).ToList();
+    List<int> faces = new();
+    List<double> vertices = new();
+
+    for (int i = 0; i < points.Count; i++)
+    {
+      if (i >= 2 && (i + 1) % 3 == 0) // every 3 new points is a new triangle
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/MultipatchFeatureToSpeckleConverter.cs:116]: similar thoughts as above comment

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/MultipatchFeatureToSpeckleConverter.cs [hunk 10]
```diff
@@ -44,83 +42,127 @@ public IReadOnlyList<Base> Convert(ACG.Multipatch target)
...
+    return new()
+    {
+      faces = faces,
+      vertices = vertices,
+      units = _settingsStore.Current.SpeckleUnits
+    };
+  }
+
+  private SOG.Mesh GetMeshFromTriangleFanPatch(List<ACG.MapPoint> points)
+  {
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/MultipatchFeatureToSpeckleConverter.cs:132]: I want to know what are they and `from what to what` we are trying to convert. mostly it is a question for @KatKatKateryna. could you document these scenarios on figma by giving some examples with shapes or real scenarios?

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/PointToSpeckleConverter.cs [hunk 11]
```diff
@@ -15,14 +14,14 @@ public PointToSpeckleConverter(IConverterSettingsStore<ArcGISConversionSettings>
...
     {
       // reproject to Active CRS
       if (
-        GeometryEngine.Instance.Project(target, _settingsStore.Current.ActiveCRSoffsetRotation.SpatialReference)
-        is not MapPoint reprojectedPt
+        ACG.GeometryEngine.Instance.Project(target, _settingsStore.Current.ActiveCRSoffsetRotation.SpatialReference)
+        is not ACG.MapPoint reprojectedPt
       )
       {
         throw new ValidationException(
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/Raw/PointToSpeckleConverter.cs:27]: i know this was pre-existing but curious what is the `ValidationException`, don't we suppose to use some conversion typed exceptions

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/TopLevel/CoreObjectsBaseToSpeckleTopLevelConverter.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    string type = target.GetType().Name;
+
+    // get display value
+    List<Base> display = _displayValueExtractor.GetDisplayValue(target).ToList();
+
+    // get properties
+    Dictionary<string, object?> properties = _propertiesExtractor.GetProperties(target);
+
+    // get application id. test for subtypes before defaulting to base type.
+    string applicationId = target is ACD.Row row
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/ToSpeckle/TopLevel/CoreObjectsBaseToSpeckleTopLevelConverter.cs:41]: conditions a bit confusing but...

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/Utils/ArcGISFieldUtils.cs [hunk 13]
```diff
@@ -80,14 +80,14 @@ public RowBuffer AssignFieldValuesToRow(
     return rowBuffer;
   }
 
-  public List<FieldDescription> GetFieldsFromSpeckleLayer(GisLayer target)
+  public List<FieldDescription> GetFieldsFromSpeckleLayer(Collection target)
   {
-    if (target["attributes"] is Base attributes)
+    if (target["fields"] is Dictionary<string, string> attributes)
```
Comment (oguzhankoral) [inline Converters/ArcGIS/Speckle.Converters.ArcGIS3/Utils/ArcGISFieldUtils.cs:85]: do we attach properties to collections in arcgis?

**Review by oguzhankoral — CHANGES_REQUESTED** [review]
- Few change requests
- Mostly questions

@clairekuang happy to do live PR review together over my comments

File: Connectors/ArcGIS/Speckle.Connectors.ArcGIS3/HostApp/ArcGISLayerUnpacker.cs [hunk 14]
```diff
@@ -1,209 +1,92 @@
...
+        collection["mapLayerType"] = layer.MapLayerType.ToString();
+        break;
+
+      case ADM.ILayerContainer:
+      default:
+        break;
     }
 
-    return gisLayer;
+    parentCollection.elements.Add(collection);
```
[the remark this answers is not in the document]
Comment (clairekuang) [reply Connectors/ArcGIS/Speckle.Connectors.ArcGIS3/HostApp/ArcGISLayerUnpacker.cs:87]: done

File: Connectors/CSi/Speckle.Connectors.CSiShared/Speckle.Connectors.CSiShared.projitems [hunk 15]
```diff
@@ -16,14 +16,12 @@
     <Compile Include="$(MSBuildThisFileDirectory)HostApp\CSiSendCollectionManager.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Operations\Send\CSiRootObjectBuilder.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Plugin\CSiPluginBase.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)Plugin\SpeckleFormBase.cs">
-      <SubType>Form</SubType>
```
[the remark this answers is not in the document]
Comment (clairekuang) [reply Connectors/CSi/Speckle.Connectors.CSiShared/Speckle.Connectors.CSiShared.projitems:20]: will remove all csi related changes

File: Converters/ArcGIS/Speckle.Converters.ArcGIS3/Utils/ArcGISFieldUtils.cs (hunk 13 above)
Comment (clairekuang) [reply Converters/ArcGIS/Speckle.Converters.ArcGIS3/Utils/ArcGISFieldUtils.cs:85]: yes, some applications like arcgis and navis have meaningful properties on collections (not jsut objects) - we don't have a convention for handling this, could be some `DataCollection: IDataCollection:IProperties` class in the future

**Review by oguzhankoral — APPROVED** [review]

**Review by KatKatKateryna — CHANGES_REQUESTED** [review]
- Color Proxies are sending wrong colors for some objects: https://latest.speckle.systems/projects/2af60ce1b6/models/f537cf39e4@cc67e77a17 
image
and https://latest.speckle.systems/projects/2af60ce1b6/models/f537cf39e4@ccff19500a
image

- Error without clear stack trace when sending group layers: 
image

**Pull request merged**